### DISCUSSION TITLE
Board definitions

### DISCUSSION THREAD

**Discussion by ropg** [ropg/heltec_esp32_lora_v3#14] [General]
# IGNORE EVERYTHING IN THIS THREAD

### HERE FOR HISTORICAL REASONS ONLY...

**EDIT: I changed board definitions again, reverting to using Heltec's own definition. Making my own is just too much of a maintenance nightmare: the Arduino ecosystem is not built for alternative board definitions and working around that is not worth it in this case. The README has been changed accordingly. There will be a separate HOWTO for using different partitioning of the flash for those that needed that feature from my own definitions.**

Hi all,

I'm thinking of changing how the board definitions are added, because having my own board manager URL – while it seemed such a cool idea at the time – is a major can of worms. The intricacies of the Arduino Board Manager process would force me to stay up on way to much arcane stuff that I don't want to have to keep updating. I think I have a better way of doing this, that is less complex and hopefully not as shaky as the current method. Sadly a bit more involved on the user end, but can't win them all.

If you've been having trouble, please help me test this before I add it to the main README and remove the other method.

<hr>

## 0. Clean up a bit

* Start Arduino IDE and remove the Heltec Unofficial URL as well as any ESP32 URLs from the additional board manager URLs field in the preferences.
* Stop Arduino IDE
* Find your Arduino library directory. On my Mac, it lives at `~/Library/Arduino15/`. There, delete boards.json and from the  packages subdir delete everything related to the heltec unofficial package
* You may also want to delete anything related to ESP32 so you're sure to start with a clean slate.  

## 1. Make sure you have the latest version of this library.

On the Mac this would be:

```
cd ~/Documents/Arduino/libraries
rm -rf Heltec_ESP32_LoRa_v3
git clone https://github.com/ropg/Heltec_ESP32_LoRa_v3
```

## 2. Install esp32 from Espressif

* Start Arduino IDE and add the following URL to the additional board manager URLs field in Preferences: `https://raw.githubusercontent.com/espressif/arduino-esp32/gh-pages/package_esp32_index.json`
* Go to board manager and install "esp32 by Espressif Systems". At this time, the latest version is 2.0.15.
* Stop Arduino IDE

## 3. Copy files

* Copy the files for the board definitions. The variants subdirectories should just get copied to the variants directory in the board manager directory, which on my Mac lives in `~/Library`. The `boards-ht_u.txt` needs to be APPENDED to the existing `boards.txt` file. If you're not on a Mac, change these directories to where your files live. (I will naturally make this more clear in the eventual instructions in the README.)

```
from=~/Documents/Arduino/libraries/Heltec_ESP32_LoRa_v3/boards
to=~/Library/Arduino15/packages/esp32/hardware/esp32/2.0.15
cp -r $from/variants/* $to/variants/
cat $from/boards-ht_u.txt >> $to/boards.txt
```

## 4. Use

* Start Arduino IDE and from the board dropdown select "Select another board and port" and enter "ht_u" in the search box. My three board definitions, whose names now start with ht_u (Heltec Unofficial) should show up.

&nbsp;

To anyone that's been having issues: please let me know if this works for you.

I'm also thinking of changing the name of the main heltec.h file to heltec_unofficial.h so the library can coexist with the heltec libraries made by Heltec.

**Comment by tao480** [comment]
Sounds like a good approach Rop,

Unfortunately I won't have time to test it since I've got a deadline for some of my own projects.  

I particularly like your suggestion of changing the name of the main heltec.h file to heltec_unofficial.h.

Many thanks for your excellent work to clean up the Heltec "soup".

**Comment by mludlum** [comment]
I was able to test this.  Seems to work.   Thanks for the quick response.  I had a lot of trouble with the Heltec libraries, but I like the hardware.   Let me know if I can contribute at all.  

I tried a couple of your examples to make sure things compiled and uploaded.  I still had to install HotButton and TimeLib, per another post here. 

Here some notes for windows users: 

~/Library/Arduino15/  ---->   c:\Users\<user>\AppData\Local\Arduino15
~/Documents/Arduino/libraries  ---->  c:\Users\<user>\Documents\Arduino\libraries
xcopy /s c:\Users\<user>\Documents\Arduino\libraries\Heltec_ESP32_LoRa_v3\boards\variants C:\Users\Michael\AppData\Local\Arduino15\packages\esp32\hardware\esp32\2.0.15\variants

**Reply by ropg** [reply]
HotButton is listed as a dependency, so that should come automatically when the library is installed using the library manager. Why TimeLib?

**Reply by mludlum** [reply]
After cleaning out all Arduino files, I started with a fresh install and followed your instructions above.  Then I selected examples from File->Examples-Heltec_Esp32_Lora_v3.   First one was Lora_rx_tx.  That would not compile unless I manually installed HotButton.  The second one was under display demos->clock demo.  That was importing TimeLib.h, so I installed the Time library.  

I don't know how examples work and if they are supposed to prompt to install for dependencies or if a user has to figure that out on their own.  Just mentioning it.

**Reply by ropg** [reply]
Thanks for testing!

The dependency (HotButton) probably did not get installed because you just got the library from GitHub (as per the test instructions) and not via the library manager.

The clock example indeed uses the time library. Says in the source, I think, will check. That came with the display library, of which I do not need to carry a fork anymore since they took all my changes. Will make that a dependency soon.

**Comment by gradew** [comment]
I can confirm that adding https://raw.githubusercontent.com/espressif/arduino-esp32/gh-pages/package_esp32_index.json was all it took to get it working on the latest Arduino IDE (2.3.2). Thanks for this :)

**Comment by tmmlrd** [comment]
I installed esp32 from Espressif as per the instructions above (point 2) but I don't have any folder /boards under ~/Documents/Arduino/libraries/Heltec_ESP32_LoRa_v3.
I have only the folders /examples /images and /src.
Can you please help me?

P.S. Before getting to you I tried the Heltec libraries and boards without success (impossibile to compile any example).

**Reply by ropg** [reply]
But you do find the `~/Documents/Arduino/libraries/Heltec_ESP32_LoRa_v3` directory itself?

If so, you haven't installed the latest version of this library. If not, your Arduino sketched (and this the libraries sub-directory) may be installed in a non-standard directory.

**Reply by tmmlrd** [reply]
For some unknown reason I had version 0.5.0 of the library. Strange since I installed it just yesterday through Arduino IDE... 
Just updated to 0.8.0 and now seems everything is working as per your directions.
Thanks a lot!

**Reply by tmmlrd** [reply]
P.S: I am not able to make the OLED display to work... tried the example SSD1306SimpleDemo: it compiles and uploads smoothly but the OLED remains off. Forced SDA and SCL to 19 and 18 and also to SDA_OLED and SCL_OLED but no change...

**Reply by ropg** [reply]
Which of the three boards do you have?

**Reply by ropg** [reply]
Also you say SSD1306SimpleDemo, which is the name as an example for the display library. I would use the version of the example from my library, where it's under File / Examples / Heltec ESP#2 LoRa v3 / display demos / SimpleDemo.

**Reply by tmmlrd** [reply]
Tried the sketch you suggested from your library (I dind't spot it before) but still nothing on the display. Compilation and upload ok (see below the report), but the display remained off. OLED display was working ok when the board (ESP32 WiFi LoRa v3) was brand new, then it stopped working when I uploaded different WiFi and Web Server examples (which is expected).

_Sketch uses 355237 bytes (10%) of program storage space. Maximum is 3342336 bytes.
Global variables use 20400 bytes (6%) of dynamic memory, leaving 307280 bytes for local variables. Maximum is 327680 bytes.
esptool.py v4.5.1
Serial port COM8
Connecting....
Chip is ESP32-S3 (revision v0.2)
Features: WiFi, BLE
Crystal is 40MHz
MAC: 64:e8:33:66:6b:74
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 921600
Changed.
Configuring flash size...
Flash will be erased from 0x00000000 to 0x00003fff...
Flash will be erased from 0x00008000 to 0x00008fff...
Flash will be erased from 0x0000e000 to 0x0000ffff...
Flash will be erased from 0x00010000 to 0x00066fff...
Compressed 15104 bytes to 10430...
Writing at 0x00000000... (100 %)
Wrote 15104 bytes (10430 compressed) at 0x00000000 in 0.3 seconds (effective 402.6 kbit/s)...
Hash of data verified.
Compressed 3072 bytes to 146...
Writing at 0x00008000... (100 %)
Wrote 3072 bytes (146 compressed) at 0x00008000 in 0.0 seconds (effective 741.4 kbit/s)...
Hash of data verified.
Compressed 8192 bytes to 47...
Writing at 0x0000e000... (100 %)
Wrote 8192 bytes (47 compressed) at 0x0000e000 in 0.1 seconds (effective 750.1 kbit/s)...
Hash of data verified.
Compressed 355600 bytes to 188635...
Writing at 0x00010000... (8 %)
Writing at 0x0001aeb4... (16 %)
Writing at 0x0002c9d7... (25 %)
Writing at 0x0003208a... (33 %)
Writing at 0x0003780a... (41 %)
Writing at 0x0003d596... (50 %)
Writing at 0x00042a38... (58 %)
Writing at 0x000480d1... (66 %)
Writing at 0x0004d702... (75 %)
Writing at 0x00055fd1... (83 %)
Writing at 0x0005e236... (91 %)
Writing at 0x00063bd3... (100 %)
Wrote 355600 bytes (188635 compressed) at 0x00010000 in 2.9 seconds (effective 973.0 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin..._

**Comment by ropg** [comment]
There was a bug in display demos/SimpleDemo, I fixed it by adding `heltec_setup();` to the `void setup()` function.

The other display demos work. Here at least. If DrawingDemo doesn't work for you, it's likely that you are having a hardware problem in addition to there being a bug in SimpleDemo.

**Reply by tmmlrd** [reply]
It finally works! Thanks a lot, I think I will make use of your libraries for my next projects with the Heltec board.

**Reply by ropg** [reply]
Yes, please do. And please post a topic here about things you make.

**Comment by tmmlrd** [comment]
Sorry, I have what seems a problem of compatibility of your library with other libraries, in particular when I introduce a both.print in my sketch the serial monitor keeps on scrolling the following message:

_Rebooting...
 ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0xc (RTC_SW_CPU_RST),boot:0x29 (SPI_FAST_FLASH_BOOT)
Saved PC:0x4202a422
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3808,len:0x4bc
load:0x403c9700,len:0xbd8
load:0x403cc700,len:0x2a0c
entry 0x403c98d0
BME280 test
-- Default Test --

TGuru Meditation Error: Core  1 panic'ed (StoreProhibited). Exception was unhandled.

Core  1 register dump:
PC      : 0x400570e8  PS      : 0x00060230  A0      : 0x82005484  A1      : 0x3fcebd10  
A2      : 0x00000000  A3      : 0x00000000  A4      : 0x00000400  A5      : 0x00000000  
A6      : 0x3fcec154  A7      : 0x00000040  A8      : 0x803804a8  A9      : 0x3fcebce0  
A10     : 0x3fcec830  A11     : 0x3fcec830  A12     : 0x00000000  A13     : 0x00000001  
A14     : 0x00060e20  A15     : 0x00000001  SAR     : 0x0000001e  EXCCAUSE: 0x0000001d  
EXCVADDR: 0x00000000  LBEG    : 0x400570e8  LEND    : 0x400570f3  LCOUNT  : 0x0000003f  


Backtrace: 0x400570e5:0x3fcebd10 0x42005481:0x3fcebd20 0x42005659:0x3fcebd40 0x42029729:0x3fcebd60 0x42029a89:0x3fcebd80 0x42005baf:0x3fcebda0 0x42005bc1:0x3fcebdc0 0x42001cf3:0x3fcebde0 0x42001d7b:0x3fcebe00 0x42008431:0x3fcebe20




ELF file SHA256: b4ef00316f8e35c5_

In my sketch to read a BME280 sensor I have included the following libraries 

#include <Wire.h>
#include <Adafruit_Sensor.h>
#include <Adafruit_BME280.h>
#include <heltec_unofficial.h>

Everything works nominally when I replace the both.print with Serial.print in my sketch.

Could you please help? 

P.S. upgraded your library today.

**Reply by ropg** [reply]
What kind of hardware do you have? Regular or Wireless Stick?

**Reply by tmmlrd** [reply]
It is a ESP32 WiFi LoRa V3 board. BTW I am using pin GPIO 4, 6, 3 and 5 for the SPI to the BME180 board, as far as I understand from the schematics the OLED uses GPIO 17, 18 and 21 so it should not be a conflict between the pin used by the two devices (that was my first hypothesis).

**Reply by ropg** [reply]
Can you show the code where you set up the SPI for the BME180?

**Reply by tmmlrd** [reply]
My sketch pasted here below:

```cpp
/*********
  Complete project details at https://randomnerdtutorials.com  
*********/

/*
Prima versione funziona se non uso both.print, altrimenti da un errore. Scritto messaggio su github allo 
sviluppatore della libreria

Versione 2: provo ad implementare il deep sleep mode, che interrompo solo per leggere il sensore e stampare

*/

#include <Wire.h>
#include <Adafruit_Sensor.h>
#include <Adafruit_BME280.h>
#include <heltec_unofficial.h>

#include <SPI.h>
#define BME_SCK 4  //<--> SCL on BME180
#define BME_MISO 6 //<-->SDO on BME180
#define BME_MOSI 3 //<-->SDA on BME180
#define BME_CS 5  //<-->CSB on BME180

#define SEALEVELPRESSURE_HPA (1013.25)

//Adafruit_BME280 bme; // I2C
//Adafruit_BME280 bme(3, 4); ///provo questa vedi https://forum.arduino.cc/t/heltec-lora-esp32-w-oled/701141/4
//Adafruit_BME280 bme(BME_CS); // hardware SPI
Adafruit_BME280 bme(BME_CS, BME_MOSI, BME_MISO, BME_SCK); // software SPI

unsigned long delayTime;

void setup() {
  Serial.begin(115200);
  Serial.println(F("BME280 test"));

  bool status;

  // default settings
  // (you can also pass in a Wire library object like &Wire2)
  status = bme.begin(0x76);  
  if (!status) {
    Serial.println("Could not find a valid BME280 sensor, check wiring!");
    while (1);
  }

  Serial.println("-- Default Test --");
  delayTime = 1000;

  Serial.println();
}


void loop() { 
  printValues();
  delay(delayTime);
}

void printValues() {
  Serial.print("Temperature = ");
  Serial.print(bme.readTemperature());
  Serial.println(" *C");
  
  // Convert temperature to Fahrenheit
  /*Serial.print("Temperature = ");
  Serial.print(1.8 * bme.readTemperature() + 32);
  Serial.println(" *F");*/
  
  Serial.print("Pressure = ");
  Serial.print(bme.readPressure() / 100.0F);
  Serial.println(" hPa");

  //Serial.print("Approx. Altitude = ");
  //Serial.print(bme.readAltitude(SEALEVELPRESSURE_HPA));
  //Serial.println(" m");

  Serial.print("Humidity = ");
  Serial.print(bme.readHumidity());
  Serial.println(" %");

  Serial.println();
}
```

**Comment by ropg** [comment]
And so this doesn't work with `both.print`?

Are you then also putting `heltec_setup()` in `setup()` ? And when you put that in, you can leave out `Serial.begin(115200)`

**Reply by tmmlrd** [reply]
The above sketch works with Serial.print but not with both.print. 

I'll try heltec_setup() as soon and let you know.

**Reply by ropg** [reply]
Without `heltec_setup()` in your `setup()`, the screen does not get initialized and `both.print` (or `display.print`) will not work, so I fully suspect this fixes your problem.

**Reply by tmmlrd** [reply]
It worked perfectly fine. Thank you!

BTW: I see the OLED scrolls whenever a new line is added. How can I clear the display, so the position of the new line does not depend on the past scrolling history?

**Reply by ropg** [reply]
`display.cls()`

Says so in the [documentation](https://github.com/ThingPulse/esp8266-oled-ssd1306?tab=readme-ov-file#arduino-print-functionality)  ;-)

**Reply by carlomigliorini** [reply]
Hi Ropg, sorry to bother you, but following STRICTLY your indication for the latest version of your precious library using an Heltec Wireless Stick V3, I get the runtime error :  E (44) i2c: i2c_set_pin(882): scl and sda gpio numbers are the same 
It seems that the board definition is wrong. Any suggestion ?

**Reply by ropg** [reply]
@carlomigliorini What versions of ESP32 Arduino and Arduino IDE are you using? What board name did you select?

**Reply by tmmlrd** [reply]
I had the same issue yesterday after an update of some libraries. I suspect it is the OLED libraries from Adafruit (if I recall...).
I hadn't time to investigate so far.

BTW is there a log or history of the library updates in Arduino IDE? I always accept all the updates proposed but in this case I may want to get back, but I am not sure which libraries was impacted but the yesterday update.

**Comment by carlomigliorini** [comment]
Hi Ropg,

- Arduino IDE 2.3.2 reinstalled after environment cleanup (e.g. delete of library directory under Documents/Arduino/library etc.)

- Phisical Board Heltec Wireless Stick v3

- Board selected in Arduino : Heltec WiFi Lora32 (V3) / Wireless shell(v3)/...

 

put inside the source file:

#define HELTEC_WIRELESS_STICK

#include <heltec_unofficial.h>
 
I should have done everything according to your last explanation contained in the readme file at the link
https://github.com/ropg/heltec_esp32_lora_v3
 
BR
Carlo Migliorini

**Reply by tmmlrd** [reply]
I confirm that after a new update of the libraries as suggested at the Arduino IDE startup everything is working nominal including BME280 (SPI interface, not I2C since I didn't manage to have that working...) and OLED. From what I understand it is this library https://github.com/ThingPulse/esp8266-oled-ssd1306 which contains the fix in version 4.6.1
